### ISSUE TITLE
🐛 Bug Report: 413 when too many entities are assigned

### ISSUE THREAD

**Issue by henrysachs** [backstage/backstage#20188]
### 📜 Description

When a user is owner of too many entities (in my case groups) the corresponding JWT Token is growing too big and the browser is rejecting these Requests with a 413. 

### 👍 Expected behavior

TBH I'm not quite sure what I'm expecting. Maybe some Docs on how to increase the maximum header size because Backstage itself can't mitigate this.

### 👎 Actual Behavior with Screenshots

Chrome DevTools show that each request is rejected with a 413 (Request headers too large)  Will add screenshots after creating of the issue

### 👟 Reproduction steps

I had a custom gitlab instance and used the [gpt](https://gitlab.com/gitlab-org/quality/performance) from GitLab to create 1000 Groups. What I didn't think about was that my user is now owner of all of them and the [GitLab Discovery](https://backstage.io/docs/integrations/gitlab/discovery/) is creating Backstage groups for all of them. Probably someone don't need GitLab to reproduce this and could just create 1000 Group Entities via a `Groups.yaml`. 

### 📃 Provide the context for the Bug.

I wasn't able to use backstage anymore as any request was failing

### 🖥️ Your Environment

```
OS:   Darwin 22.6.0 - darwin/arm64
node: v18.18.0
yarn: 3.6.1
cli:  0.22.13 (installed)
backstage:  1.18.1

Dependencies:
  @backstage/app-defaults                                          1.4.3
  @backstage/backend-app-api                                       0.5.3
  @backstage/backend-common                                        0.19.5
  @backstage/backend-defaults                                      0.2.3
  @backstage/backend-dev-utils                                     0.1.1
  @backstage/backend-openapi-utils                                 0.0.4
  @backstage/backend-plugin-api                                    0.6.3
  @backstage/backend-tasks                                         0.5.8
  @backstage/catalog-client                                        1.4.4
  @backstage/catalog-model                                         1.4.2
  @backstage/cli-common                                            0.1.12
  @backstage/cli-node                                              0.1.4
  @backstage/cli                                                   0.22.13
  @backstage/config-loader                                         1.5.0
  @backstage/config                                                1.1.0
  @backstage/core-app-api                                          1.10.0
  @backstage/core-components                                       0.13.5
  @backstage/core-plugin-api                                       1.6.0
  @backstage/dev-utils                                             1.0.21
  @backstage/errors                                                1.2.2
  @backstage/eslint-plugin                                         0.1.3
  @backstage/integration-aws-node                                  0.1.6
  @backstage/integration-react                                     1.1.19
  @backstage/integration                                           1.7.0
  @backstage/plugin-api-docs                                       0.9.11
  @backstage/plugin-app-backend                                    0.3.51
  @backstage/plugin-app-node                                       0.1.3
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.1.0
  @backstage/plugin-auth-backend-module-github-provider            0.1.0
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.0
  @backstage/plugin-auth-backend-module-google-provider            0.1.0
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.0
  @backstage/plugin-auth-backend                                   0.19.0
  @backstage/plugin-auth-node                                      0.3.0
  @backstage/plugin-catalog-backend-module-gitlab                  0.3.0
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.0
  @backstage/plugin-catalog-backend-module-unprocessed             0.3.0
  @backstage/plugin-catalog-backend                                1.13.0
  @backstage/plugin-catalog-common                                 1.0.16
  @backstage/plugin-catalog-graph                                  0.2.36
  @backstage/plugin-catalog-import                                 0.10.0
  @backstage/plugin-catalog-node                                   1.4.4
  @backstage/plugin-catalog-react                                  1.8.4
  @backstage/plugin-catalog                                        1.13.0
  @backstage/plugin-events-node                                    0.2.12
  @backstage/plugin-explore-common                                 0.0.2
  @backstage/plugin-github-actions                                 0.6.5
  @backstage/plugin-home-react                                     0.1.3
  @backstage/plugin-home                                           0.5.8
  @backstage/plugin-kubernetes-backend                             0.12.0
  @backstage/plugin-kubernetes-common                              0.6.6
  @backstage/plugin-kubernetes                                     0.10.3
  @backstage/plugin-org                                            0.6.14
  @backstage/plugin-permission-backend                             0.5.26
  @backstage/plugin-permission-common                              0.7.8
  @backstage/plugin-permission-node                                0.7.14
  @backstage/plugin-permission-react                               0.4.15
  @backstage/plugin-proxy-backend                                  0.4.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.2.6
  @backstage/plugin-scaffolder-backend                             1.17.0
  @backstage/plugin-scaffolder-common                              1.4.1
  @backstage/plugin-scaffolder-node                                0.2.3
  @backstage/plugin-scaffolder-react                               1.5.5
  @backstage/plugin-scaffolder                                     1.15.0
  @backstage/plugin-search-backend-module-catalog                  0.1.7
  @backstage/plugin-search-backend-module-explore                  0.1.7
  @backstage/plugin-search-backend-module-pg                       0.5.12
  @backstage/plugin-search-backend-module-techdocs                 0.1.7
  @backstage/plugin-search-backend-node                            1.2.7
  @backstage/plugin-search-backend                                 1.4.3
  @backstage/plugin-search-common                                  1.2.6
  @backstage/plugin-search-react                                   1.7.0
  @backstage/plugin-search                                         1.4.0
  @backstage/plugin-tech-radar                                     0.6.8
  @backstage/plugin-techdocs-backend                               1.7.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.0
  @backstage/plugin-techdocs-node                                  1.8.0
  @backstage/plugin-techdocs-react                                 1.1.11
  @backstage/plugin-techdocs                                       1.7.0
  @backstage/plugin-user-settings                                  0.7.10
  @backstage/release-manifests                                     0.0.10
  @backstage/test-utils                                            1.4.3
  @backstage/theme                                                 0.4.2
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.5

``
`

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by awanlin** [comment]
Hi @henrysachs, I don't have an answer to this but there was a recent thread on the Backstage Discord server that might be worth watching as well: https://discord.com/channels/687207715902193673/1151635622671417354

@kuangp were you able to resolve your issue? Seems very similar to this one 🤔

**Comment by kuangp** [comment]
hm from what I can find according to https://stackoverflow.com/questions/686217/maximum-on-http-header-values there is no standard that limits the size of HTTP headers so this restriction seems to be coming from your server side and not the browser (which was the issue with cookie size limits in my issue)

If this is hitting a standard Backstage backend instance which is utilizing NodeJs + Express.js currently then it looks like this may be a potential solution for you: https://github.com/expressjs/express/issues/4131

**Comment by Rugvip** [comment]
I'd back up a bit and say that you should not be including 1000s of groups in our ownership refs. The ownership refs are only the groups that you claim ownership through, as in if group `team-a` owns the `playlists` service, and you list `team-a` in your ownership refs then you are considered to be directly responsible for and own the `playlists` service. Being directly responsible for all services for 1000s of groups is not feasible 😅. More on the ownership refs over here too: https://backstage.io/docs/auth/identity-resolver#backstage-user-identity

Slim down the number of listed ownership refs somehow, either by customizing the sign-in resolver logic or by making sure you are not a member of that many groups. If you do want some form of admin permission implemented, that's better implemented in a custom permission resolver.

**Comment by henrysachs** [comment]
Hey There, sorry for the late reply. First of all @Rugvip I'm totally on the side of "you shouldn't own 1000 groups" this was more an accident when using the gpt tool and noticing that the Token was growing and growing when owning more entities. So my question is more if it should be documented how to increase the header size. I also found the express Issue but we start our backend with the `backstage-cli start` command so I have no Idea how to pass "node.js" arguments to that.

**Comment by Rugvip** [comment]
Tbh what we might want to do here is to limit the number of ownership refs in issued tokens to some high but sensible value, just to catch mistakes. Perhaps something like limiting to 100 refs

**Comment by henrysachs** [comment]
This is the kind of discussion I wanted to start. Like should backstage itself limit this or should there be an Entry in the documentation that this can happen and you can (if you had some serious thoughts about it) increase the max header value with this command line flag.

**Comment by henrysachs** [comment]
@Rugvip how do we want to go further on this one?

**Comment by henrysachs** [comment]
yeah i would just like to limit it in backstage because i thought this might lead to unexpected errors or vectors

**Comment by freben** [comment]
For anyone tracking this:

As part of the large auth improvements work that's underway right now, we're moving the ownership refs out of the tokens and exposing them via a userInfo endpoint instead for those who need it. This will make the tokens fixed size and fix this issue once and for all. But it'll be a bit longer until that lands.

**Comment by benjdlambert** [comment]
I think that most of the `userInfo` work has now been completed right @freben? Does this mean that this should be fixed now? 🙏

**Comment by freben** [comment]
Not quite ready yet - the Backstage tokens do contain `ent` claims for a bit longer, while the limited user tokens (cookies) exclude them in recent releases.

**Comment by benjdlambert** [comment]
Think this got closed inadvertently! let's come back to this now we've shipped half a fix for the cookie splitting.

**Comment by RobinWeitzel** [comment]
Any update on this topic? I encountered the same issue.
For me, the use case is that the admins of GitLab (who own 1000s of groups) are also the admins of Backstage. This means those admins cannot log into a tool they are responsible for (while normal users can since they own a reasonable number of groups).

**Comment by freben** [comment]
Some time back, we introduced the `coreServices.userInfo` that lets services fetch all of the claims of a user (including the large ownership refs that are being discussed here).

However, we did NOT stop putting those refs also in the token at that time, because we needed to ensure that people got some time to move over to actually using that new core service instead of making assumptions about the token contents. Therefore the issue remains of tokens growing with the number of ownership refs.

We'll address that some time in the future, but no strict time frame exists right now.

Until that happens, your best bet is to make a custom sign in resolver that prunes the list of ownership refs down to a more sensible size, possibly using some heuristic that makes sense to you. For example, it's common to only put org teams that you are actually a real physical member of in there, which usually is a very very small number.

**Comment by drodil** [comment]
We also hit this with one user having 94 groups in the `ents`. That shouldn't be too much (right?), but reported also the #28836 issue that might or might not be caused by this.

**Comment by vinzscam** [comment]
I think it's a duplicate of https://github.com/backstage/backstage/issues/23303

**Comment by DB-Vincent** [comment]
We are facing the same issue as @RobinWeitzel, any pointers on how we can check which groups we should actually be assign to an admin user (at least as an intermediary solution until @freben and the team can pick this issue up)?

**Comment by freben** [comment]
I'll mention that in last Tuesday's `-next` release, [a flag was added](https://github.com/backstage/backstage/blob/master/docs/releases/v1.39.0-next.1-changelog.md#patch-changes-40) that lets you opt in to exclude the large claim from your tokens.

Even after doing that, there are some cases where you can still hit limits. For example on the regular `/catalog` page if you have Owned selected, the size of the outgoing requests (filter complexity) scale with the number of refs too. There's still a ton of value in keeping the number of refs as low as possible, both for performance and security and reliability reasons.

**Comment by MihaiTabara** [comment]
Un-stale - still ongoing and work-in-progress in related PRs.

**Comment by drodil** [comment]
Not stale, still waiting for review for https://github.com/backstage/backstage/pull/29983

**Comment by freben** [comment]
@drodil Does that really address the issue? I thought this issue was more related to [making `omitIdentityTokenOwnershipClaim` default](https://github.com/backstage/backstage/blob/03fcaf3666bfddf791fb1f6f67e33e315d649cea/plugins/auth-backend/config.d.ts#L53) true for people.

**Comment by drodil** [comment]
@freben Ah, true, the 413 with permissions is another issue when a user owns too many entities, leading to too large requests to the permission framework. But a bit related :D

**Comment by elaine-mattos** [comment]
As information for future readers, the links below might be useful :)

https://github.com/backstage/backstage/pull/32874
https://backstage.io/docs/features/software-catalog/api/query-entities-by-predicate
